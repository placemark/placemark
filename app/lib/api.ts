import { PersistenceMetadataPersisted } from "./persistence/ipersistence";
import { env } from "app/lib/env_client";
import { IWrappedFeature } from "types";

function mapBaseURLWithoutSlash(metadata: PersistenceMetadataPersisted) {
  return `${env.NEXT_PUBLIC_API_BASE_WITH_SLASH}api/v1/map/${metadata.id}`;
}

/**
 * Get the URL of the feature collection API
 */
export function getAPIURL(metadata: PersistenceMetadataPersisted) {
  return `${mapBaseURLWithoutSlash(metadata)}/featurecollection`;
}

/**
 * Get the URL of a shared map
 */
export function getShareURL(metadata: PersistenceMetadataPersisted) {
  return `${env.NEXT_PUBLIC_DOMAIN_WITH_SLASH}public/${metadata.id}`;
}

/**
 * Get the URL of an individual feature within a map.
 */
export function getAPIURLFeature(
  metadata: PersistenceMetadataPersisted,
  featureId: IWrappedFeature["id"]
) {
  return `${mapBaseURLWithoutSlash(metadata)}/feature/${featureId}`;
}

export function gistUrlFromId(id: string) {
  return `https://gist.github.com/${id}`;
}
