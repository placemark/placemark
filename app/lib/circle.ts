import { Polygon } from "geojson";
import clamp from "lodash/clamp";
import { z } from "zod";
import { CIRCLE_TYPE } from "state/mode";
import { Feature } from "types";

const D2R = Math.PI / 180;
const MAXEXTENT = 20037508.342789244;
const R2D = 180 / Math.PI;
const A = 6378137.0;

export const CircleProp = z
  .object({
    "@circle": z.object({
      type: z.nativeEnum(CIRCLE_TYPE),
      center: z.tuple([z.number(), z.number()]),
    }),
  })
  .passthrough();

export type ICircleProp = z.infer<typeof CircleProp>;

function sign(x: number) {
  return x < 0 ? -1 : x > 0 ? 1 : 0;
}

/**
 * From TURF
 * Convert lon/lat values to 900913 x/y.
 * (from https://github.com/mapbox/sphericalmercator)
 */
function toMercator(position: Pos2): Pos2 {
  const [longitude, latitude] = position;

  // compensate longitudes passing the 180th meridian
  // from https://github.com/proj4js/proj4js/blob/master/lib/common/adjust_lon.js
  const adjusted =
    Math.abs(longitude) <= 180 ? longitude : longitude - sign(longitude) * 360;

  return [
    clamp(A * adjusted * D2R, -MAXEXTENT, MAXEXTENT),
    clamp(
      A * Math.log(Math.tan(Math.PI * 0.25 + 0.5 * latitude * D2R)),
      -MAXEXTENT,
      MAXEXTENT
    ),
  ];
}

/**
 * Convert 900913 x/y values to lon/lat.
 * (from https://github.com/mapbox/sphericalmercator)
 */
function toLonLat(xy: Pos2): Pos2 {
  return [
    (xy[0] * R2D) / A,
    (Math.PI * 0.5 - 2.0 * Math.atan(Math.exp(-xy[1] / A))) * R2D,
  ];
}

/**
 * Same as Turf.
 */
function degreesToRadians(degrees: number): number {
  const radians = degrees % 360;
  return (radians * Math.PI) / 180;
}

/**
 * Also same as turf.
 */
function radiansToDegrees(radians: number): number {
  const degrees = radians % (2 * Math.PI);
  return (degrees * 180) / Math.PI;
}

function getVerticesDegrees(
  center: Pos2,
  degrees: number,
  angles: number[]
): Pos2[] {
  return angles.map((bearingRad): Pos2 => {
    return [
      center[0] + Math.cos(bearingRad) * degrees,
      center[1] + Math.sin(bearingRad) * degrees,
    ];
  });
}

function getVerticesGeodesic(
  center: Pos2,
  radians: number,
  angles: number[]
): Pos2[] {
  const [longitude1, latitude1] = [
    degreesToRadians(center[0]),
    degreesToRadians(center[1]),
  ];

  return angles.map((bearingRad): Pos2 => {
    const latitude2 = Math.asin(
      Math.sin(latitude1) * Math.cos(radians) +
        Math.cos(latitude1) * Math.sin(radians) * Math.cos(bearingRad)
    );
    const longitude2 =
      longitude1 +
      Math.atan2(
        Math.sin(bearingRad) * Math.sin(radians) * Math.cos(latitude1),
        Math.cos(radians) - Math.sin(latitude1) * Math.sin(latitude2)
      );
    return [radiansToDegrees(longitude2), radiansToDegrees(latitude2)];
  });
}

/**
 * Haversine distance. Adapted from Turf.
 */
function distanceInRadians(a: Pos2, b: Pos2) {
  const dLat = degreesToRadians(b[1] - a[1]);
  const dLon = degreesToRadians(b[0] - a[0]);
  const lat1 = degreesToRadians(a[1]);
  const lat2 = degreesToRadians(b[1]);

  const ax =
    Math.pow(Math.sin(dLat / 2), 2) +
    Math.pow(Math.sin(dLon / 2), 2) * Math.cos(lat1) * Math.cos(lat2);

  return 2 * Math.atan2(Math.sqrt(ax), Math.sqrt(1 - ax));
}

function getVerticesMercator(
  center: Pos2,
  meters: number,
  angles: number[]
): Pos2[] {
  const centerMercator = toMercator(center);

  return angles.map((bearingRad): Pos2 => {
    return toLonLat([
      centerMercator[0] + Math.cos(bearingRad) * meters,
      centerMercator[1] + Math.sin(bearingRad) * meters,
    ]);
  });
}

/**
 * References
 * https://github.com/Turfjs/turf/blob/master/packages/turf-circle/index.ts
 * https://github.com/henrythasler/Leaflet.Geodesic/blob/master/src/geodesic-geom.ts#L240
 */
export function makeCircleGeodesic(center: Pos2, radians: number): Polygon {
  // TODO: auto steps
  const steps = 100;

  const angles = Array.from({ length: steps - 1 }, (_, i) => {
    return (i / steps) * Math.PI * 2;
  }).concat([0]);

  const vertices = getVerticesGeodesic(center, radians, angles);

  return {
    type: "Polygon",
    coordinates: [vertices],
  };
}

export function makeCircleDegrees(center: Pos2, degrees: number): Polygon {
  // TODO: auto steps
  const steps = 100;

  const angles = Array.from({ length: steps - 1 }, (_, i) => {
    return (i / steps) * Math.PI * 2;
  }).concat([0]);

  const vertices = getVerticesDegrees(center, degrees, angles);

  return {
    type: "Polygon",
    coordinates: [vertices],
  };
}

export function makeCircleMercator(
  center: Pos2,
  distanceInMercatorMeters: number
): Polygon {
  // TODO: auto steps
  const steps = 100;

  const angles = Array.from({ length: steps - 1 }, (_, i) => {
    return (i / steps) * Math.PI * 2;
  }).concat([0]);

  const vertices = getVerticesMercator(
    center,
    distanceInMercatorMeters,
    angles
  );

  return {
    type: "Polygon",
    coordinates: [vertices],
  };
}

function distanceInMercatorMeters(a: Pos2, b: Pos2) {
  const am = toMercator(a);
  const bm = toMercator(b);
  return Math.sqrt(Math.pow(am[0] - bm[0], 2) + Math.pow(am[1] - bm[1], 2));
}

function distanceInDegrees(am: Pos2, bm: Pos2) {
  return Math.sqrt(Math.pow(am[0] - bm[0], 2) + Math.pow(am[1] - bm[1], 2));
}

export function makeCircleNative({
  center,
  value,
  type,
}: {
  center: Pos2;
  value: number;
  type: CIRCLE_TYPE;
}): Polygon {
  switch (type) {
    case CIRCLE_TYPE.MERCATOR: {
      return makeCircleMercator(center, value);
    }
    case CIRCLE_TYPE.GEODESIC: {
      return makeCircleGeodesic(center, value);
    }
    case CIRCLE_TYPE.DEGREES: {
      return makeCircleDegrees(center, value);
    }
  }
}

export function makeCircle({
  center,
  mouse,
  type,
}: {
  center: Pos2;
  mouse: Pos2;
  type: CIRCLE_TYPE;
}): Polygon {
  switch (type) {
    case CIRCLE_TYPE.MERCATOR: {
      return makeCircleMercator(
        center,
        distanceInMercatorMeters(center, mouse)
      );
    }
    case CIRCLE_TYPE.GEODESIC: {
      return makeCircleGeodesic(center, distanceInRadians(center, mouse));
    }
    case CIRCLE_TYPE.DEGREES: {
      return makeCircleDegrees(center, distanceInDegrees(center, mouse));
    }
  }
}

export function getCircleRadius(feature: Feature) {
  if (feature.geometry?.type !== "Polygon") return null;
  const prop = CircleProp.safeParse(feature.properties);
  if (!prop.success) return null;
  const vertex = feature.geometry.coordinates[0]?.[0] as Pos2 | undefined;
  if (vertex === undefined) return null;

  const { type, center } = prop.data["@circle"];

  switch (type) {
    case CIRCLE_TYPE.DEGREES: {
      return distanceInDegrees(center, vertex);
    }
    case CIRCLE_TYPE.MERCATOR: {
      return distanceInMercatorMeters(center, vertex);
    }
    case CIRCLE_TYPE.GEODESIC: {
      return distanceInRadians(center, vertex);
    }
  }
}

export function getCircleProp(feature: Feature) {
  if (feature.geometry?.type !== "Polygon") return null;
  const prop = CircleProp.safeParse(feature.properties);
  if (!prop.success) return null;
  return prop.data["@circle"];
}
