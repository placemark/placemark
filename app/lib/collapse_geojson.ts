import {
  Feature,
  Geometry,
  MultiPolygon,
  FeatureCollection,
  MultiPoint,
  MultiLineString,
} from "types";

/**
 * Intended to be used to shrink geojson for
 * thumbnails, this collapses geometries into multigeometries.
 */
export function collapseGeoJSON(fc: FeatureCollection): FeatureCollection {
  const multiPoint: MultiPoint = {
    type: "MultiPoint",
    coordinates: [],
  };
  const multiLineString: MultiLineString = {
    type: "MultiLineString",
    coordinates: [],
  };
  const multiPolygon: MultiPolygon = {
    type: "MultiPolygon",
    coordinates: [],
  };

  function collapseGeometry(geometry: Geometry | null) {
    if (!geometry) return;

    switch (geometry.type) {
      case "Point": {
        multiPoint.coordinates.push(geometry.coordinates);
        break;
      }
      case "Polygon": {
        multiPolygon.coordinates.push(geometry.coordinates);
        break;
      }
      case "LineString": {
        multiLineString.coordinates.push(geometry.coordinates);
        break;
      }

      case "MultiPoint": {
        multiPoint.coordinates = multiPoint.coordinates.concat(
          geometry.coordinates
        );
        break;
      }
      case "MultiPolygon": {
        multiPolygon.coordinates = multiPolygon.coordinates.concat(
          geometry.coordinates
        );
        break;
      }
      case "MultiLineString": {
        multiLineString.coordinates = multiLineString.coordinates.concat(
          geometry.coordinates
        );
        break;
      }
      case "GeometryCollection": {
        for (const g of geometry.geometries) {
          collapseGeometry(g);
        }
      }
    }
  }

  for (const feature of fc.features) {
    collapseGeometry(feature.geometry);
  }

  return {
    type: "FeatureCollection",
    features: [multiPolygon, multiPoint, multiLineString]
      .filter((geom) => geom.coordinates.length)
      .map((geom): Feature => {
        return {
          type: "Feature",
          geometry: geom,
          properties: null,
        };
      }),
  };
}
