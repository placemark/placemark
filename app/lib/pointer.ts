import { JsonValue } from "type-fest";

const hasExcape = /~/;
const escapeMatcher = /~[01]/g;

type Pointer = string | string[];

function escapeReplacer(m: string) {
  switch (m) {
    case "~1":
      return "/";
    case "~0":
      return "~";
  }
  throw new Error("Invalid tilde escape: " + m);
}

function untilde(str: string) {
  if (!hasExcape.test(str)) return str;
  return str.replace(escapeMatcher, escapeReplacer);
}

function setter(obj: any, pointer: Pointer, value: any): JsonValue {
  let part: string | number;
  let hasNextPart: boolean;

  if (pointer[1] === "constructor" && pointer[2] === "prototype") {
    return obj as JsonValue;
  }
  if (pointer[1] === "__proto__") return obj as JsonValue;

  for (let p = 1, len = pointer.length; p < len; ) {
    part = untilde(pointer[p++]);
    hasNextPart = len > p;

    if (typeof obj[part] === "undefined") {
      // support setting of /-
      if (Array.isArray(obj) && part === "-") {
        part = obj.length;
      }

      // support nested objects/array when setting values
      if (hasNextPart) {
        if (
          (pointer[p] !== "" && (pointer[p] as any) < Infinity) ||
          pointer[p] === "-"
        )
          obj[part] = [];
        else obj[part] = {};
      }
    }

    if (!hasNextPart) break;
    obj = obj[part];
  }

  const oldValue = obj[part!];
  if (value === undefined) delete obj[part!];
  else obj[part!] = value;
  return oldValue as JsonValue;
}

export function compilePointer(pointer: Pointer) {
  if (typeof pointer === "string") {
    pointer = pointer.split("/");
    if (pointer[0] === "") return pointer;
    throw new Error("Invalid JSON pointer.");
  } else if (Array.isArray(pointer)) {
    return pointer;
  }

  throw new Error("Invalid JSON pointer.");
}

export function get(obj: any, pointer: Pointer): JsonValue | undefined {
  if (typeof obj !== "object") throw new Error("Invalid input object.");
  pointer = compilePointer(pointer);
  const len = pointer.length;
  if (len === 1) return obj as JsonValue;

  for (let p = 1; p < len; ) {
    obj = obj[untilde(pointer[p++])];
    if (len === p) return obj as JsonValue;
    if (typeof obj !== "object") return undefined;
  }
}

export function clone<T>(obj: T, pointer: Pointer): T {
  if (typeof obj !== "object") throw new Error("Invalid input object.");
  pointer = compilePointer(pointer);
  const len = pointer.length;
  if (len === 1) return obj;

  const rootPointer = Object.assign({}, obj);
  let levelPointer = rootPointer as any;

  for (let p = 1; p < len - 1; ) {
    const key = untilde(pointer[p++]);
    // Clone objects and arrays
    if (Array.isArray(levelPointer[key])) {
      // eslint-disable-next-line
      levelPointer[key] = levelPointer[key].slice();
    } else if (
      levelPointer[key] !== null &&
      typeof levelPointer[key] === "object"
    ) {
      levelPointer[key] = Object.assign({}, levelPointer[key]);
    }
    levelPointer = levelPointer[key];
    if (len === p) return rootPointer;
  }
  return rootPointer;
}

export function set(obj: any, pointer: Pointer, value: any) {
  if (typeof obj !== "object") throw new Error("Invalid input object.");
  pointer = compilePointer(pointer);
  if (pointer.length === 0) throw new Error("Invalid JSON pointer for set.");
  return setter(obj, pointer, value);
}

export function compile(pointer: Pointer) {
  const compiled = compilePointer(pointer);
  return {
    get: function (object: any) {
      return get(object, compiled);
    },
    set: function (object: any, value: any) {
      return set(object, compiled, value);
    },
  };
}
