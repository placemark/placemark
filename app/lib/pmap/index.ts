import mapboxgl from "mapbox-gl";
import loadAndAugmentStyle, {
  FEATURES_SOURCE_NAME,
  EPHEMERAL_SOURCE_NAME,
} from "app/lib/load_and_augment_style";
import type {
  EphemeralEditingState,
  Sel,
  Data,
  PreviewProperty,
} from "state/jotai";
import {
  CURSOR_DEFAULT,
  DEFAULT_MAP_BOUNDS,
  emptySelection,
  LASSO_YELLOW,
  LASSO_DARK_YELLOW,
  LINE_COLORS_SELECTED_RGB,
  WHITE,
  DECK_SYNTHETIC_ID,
  DECK_LASSO_ID,
} from "app/lib/constants";
import { splitFeatureGroups } from "app/lib/pmap/split_feature_groups";
import type {
  Feature,
  IPresence,
  IFeatureCollection,
  ISymbolization,
  LayerConfigMap,
  IFeature,
  Point,
} from "types";
import { makeRectangle } from "app/lib/pmap/merge_ephemeral_state";
import { colorFromPresence } from "app/lib/color";
import type { Presence } from "@prisma/client";
import { IDMap } from "app/lib/id_mapper";
import { shallowArrayEqual } from "app/lib/utils";
import { MapboxOverlay } from "@deck.gl/mapbox/typed";
import { PolygonLayer, ScatterplotLayer } from "@deck.gl/layers/typed";

const MAP_OPTIONS: Omit<mapboxgl.MapboxOptions, "container"> = {
  style: { version: 8, layers: [], sources: {} },
  boxZoom: false,
  dragRotate: false,
  attributionControl: false,
  fadeDuration: 0,
};

const cursorSvg = (color: string) => {
  const div = document.createElement("div");
  div.style.color = color;
  div.innerHTML = `<svg width="18" height="18" viewBox="0 0 18 18" fill="none" xmlns="http://www.w3.org/2000/svg">
<path d="M7 17L1 1L17 7L10 10L7 17Z" stroke="white" fill="currentColor" stroke-width="1.5" stroke-linejoin="round"/>
</svg>
`;
  return div;
};

type ClickEvent = mapboxgl.MapMouseEvent & mapboxgl.EventData;
type MoveEvent = mapboxgl.MapboxEvent & mapboxgl.EventData;

export type PMapHandlers = {
  onClick: (e: ClickEvent) => void;
  onDoubleClick: (e: ClickEvent) => void;
  onMapMouseUp: (e: mapboxgl.MapMouseEvent) => void;
  onMapMouseMove: (e: mapboxgl.MapMouseEvent) => void;
  onMapTouchMove: (e: mapboxgl.MapTouchEvent) => void;
  onMapMouseDown: (e: mapboxgl.MapMouseEvent) => void;
  onMapTouchStart: (e: mapboxgl.MapTouchEvent) => void;
  onMoveEnd: (e: mapboxgl.MapboxEvent & mapboxgl.EventData) => void;
  onMapTouchEnd: (e: mapboxgl.MapTouchEvent) => void;
  onMove: (e: mapboxgl.MapboxEvent & mapboxgl.EventData) => void;
};

const lastValues = new WeakMap<mapboxgl.GeoJSONSource, Feature[]>();

/**
 * Memoized set data for a mapboxgl.GeoJSONSource. If
 * the same source is called with the same data,
 * it won't set.
 */
function mSetData(
  source: mapboxgl.GeoJSONSource,
  newData: Feature[],
  _label: string,
  force?: boolean
) {
  if (!shallowArrayEqual(lastValues.get(source), newData) || force) {
    source.setData({
      type: "FeatureCollection",
      features: newData,
    } as IFeatureCollection);
    lastValues.set(source, newData);
  } else {
    // console.log(
    //   "Skipped update",
    //   _label,
    //   source,
    //   newData,
    //   lastValues.get(source)
    // );
  }
}

export default class PMap {
  map: mapboxgl.Map;
  handlers: React.MutableRefObject<PMapHandlers>;
  idMap: IDMap;

  lastSelection: Sel;
  lastSelectionIds: Set<RawId>;
  lastData: Data | null;
  lastEphemeralState: EphemeralEditingState;
  lastSymbolization: ISymbolization | null;
  presenceMarkers: Map<IPresence["userId"], mapboxgl.Marker>;
  lastLayer: LayerConfigMap | null;
  lastPreviewProperty: PreviewProperty;
  overlay: MapboxOverlay;

  constructor({
    element,
    layerConfigs,
    lastPresence,
    handlers,
    previewProperty,
    symbolization,
    idMap,
    controlsCorner = "bottom-left",
  }: {
    element: HTMLDivElement;
    layerConfigs: LayerConfigMap;
    lastPresence: Presence | BBox4 | null;
    handlers: React.MutableRefObject<PMapHandlers>;
    symbolization: ISymbolization;
    previewProperty: PreviewProperty;
    idMap: IDMap;
    controlsCorner?: Parameters<mapboxgl.Map["addControl"]>[1];
  }) {
    this.idMap = idMap;
    const positionOptions = lastPresence
      ? Array.isArray(lastPresence)
        ? {
            bounds: mapboxgl.LngLatBounds.convert(lastPresence),
            fitBoundsOptions: { padding: 20 },
          }
        : {
            bounds: mapboxgl.LngLatBounds.convert([
              [lastPresence.minx, lastPresence.miny],
              [lastPresence.maxx, lastPresence.maxy],
            ]),
            pitch: lastPresence.pitch,
            bearing: lastPresence.bearing,
          }
      : {
          bounds: DEFAULT_MAP_BOUNDS as mapboxgl.LngLatBoundsLike,
        };

    const map = new mapboxgl.Map({
      container: element,
      ...MAP_OPTIONS,
      ...positionOptions,
    });

    this.overlay = new MapboxOverlay({
      interleaved: true,
      layers: [],
    });

    map.addControl(this.overlay);

    map.addControl(
      new mapboxgl.GeolocateControl({
        showUserLocation: false,
        showAccuracyCircle: false,
        positionOptions: {
          enableHighAccuracy: true,
        },
      }),
      controlsCorner
    );
    map.addControl(new mapboxgl.NavigationControl({}), controlsCorner);
    map.addControl(
      new mapboxgl.AttributionControl({
        compact: true,
      })
    );
    map.getCanvas().style.cursor = CURSOR_DEFAULT;
    map.on("click", this.onClick);
    map.on("mousedown", this.onMapMouseDown);
    map.on("mousemove", this.onMapMouseMove);
    map.on("dblclick", this.onMapDoubleClick);
    map.on("mouseup", this.onMapMouseUp);
    map.on("moveend", this.onMoveEnd);
    map.on("touchend", this.onMapTouchEnd);
    map.on("move", this.onMove);

    map.on("touchstart", this.onMapTouchStart);
    map.on("touchmove", this.onMapTouchMove);
    map.on("touchend", this.onMapTouchEnd);

    this.presenceMarkers = new Map();
    this.lastSymbolization = symbolization;

    this.lastSelection = { type: "none" };
    this.lastSelectionIds = emptySelection;
    this.lastData = null;
    this.lastEphemeralState = { type: "none" };
    this.lastLayer = null;
    this.lastPreviewProperty = null;
    this.handlers = handlers;
    this.map = map;
    void this.setStyle({
      layerConfigs,
      symbolization,
      previewProperty: previewProperty,
    });
  }

  /**
   * Handler proxies --------------------------------------
   */
  onClick = (e: LayerScopedEvent) => {
    this.handlers.current.onClick(e);
  };

  onMapMouseDown = (e: LayerScopedEvent) => {
    this.handlers.current.onMapMouseDown(e);
  };

  onMapTouchStart = (e: mapboxgl.MapTouchEvent) => {
    this.handlers.current.onMapTouchStart(e);
  };

  onMapMouseUp = (e: LayerScopedEvent) => {
    this.handlers.current.onMapMouseUp(e);
  };

  onMoveEnd = (e: MoveEvent) => {
    this.handlers.current.onMoveEnd(e);
  };

  onMapTouchEnd = (e: mapboxgl.MapTouchEvent) => {
    this.handlers.current.onMapTouchEnd(e);
  };

  onMove = (e: MoveEvent) => {
    this.handlers.current.onMove(e);
  };

  onMapMouseMove = (e: mapboxgl.MapMouseEvent) => {
    this.handlers.current.onMapMouseMove(e);
  };

  onMapTouchMove = (e: mapboxgl.MapTouchEvent) => {
    this.handlers.current.onMapTouchMove(e);
  };

  onMapDoubleClick = (e: mapboxgl.MapMouseEvent) => {
    this.handlers.current.onDoubleClick(e);
  };

  setPresences(presences: IPresence[]) {
    const ids = new Set(presences.map((p) => p.userId));
    for (const presence of presences) {
      const marker =
        this.presenceMarkers.get(presence.userId) ??
        new mapboxgl.Marker(cursorSvg(colorFromPresence(presence)));
      marker
        .setLngLat([presence.cursorLongitude, presence.cursorLatitude])
        .addTo(this.map);
      this.presenceMarkers.set(presence.userId, marker);
    }
    // Remove stale presences
    for (const [id, marker] of this.presenceMarkers.entries()) {
      if (!ids.has(id)) {
        marker.remove();
        this.presenceMarkers.delete(id);
      }
    }
  }

  /**
   * The central hard method, trying to optimize feature updates
   * on the map.
   */
  setData({
    data,
    ephemeralState,
    force = false,
  }: {
    data: Data;
    ephemeralState: EphemeralEditingState;
    force?: boolean;
  }) {
    if (!(this.map && (this.map as any).style)) {
      this.lastData = data;
      return;
    }

    const featuresSource = this.map.getSource(
      FEATURES_SOURCE_NAME
    ) as mapboxgl.GeoJSONSource;

    const ephemeralSource = this.map.getSource(
      EPHEMERAL_SOURCE_NAME
    ) as mapboxgl.GeoJSONSource;

    if (!featuresSource || !ephemeralSource) {
      // Set the lastFeatureList here
      // so that the setStyle method will
      // add it again. This happens when the map
      // is initially loaded.
      this.lastData = data;
      return;
    }

    const groups = splitFeatureGroups({
      idMap: this.idMap,
      data,
      lastSymbolization: this.lastSymbolization,
      previewProperty: this.lastPreviewProperty,
    });

    // console.log(
    //   "in setData",
    //   JSON.stringify({
    //     newSelection,
    //     outputIds: [...groups.selectionIds],
    //   })
    // );
    // TODO: fix flash
    mSetData(ephemeralSource, groups.ephemeral, "ephem");
    mSetData(featuresSource, groups.features, "features", force);

    this.overlay.setProps({
      layers: [
        new ScatterplotLayer<IFeature<Point>>({
          id: DECK_SYNTHETIC_ID,

          radiusUnits: "pixels",
          lineWidthUnits: "pixels",

          pickable: true,
          stroked: true,
          filled: true,

          data: groups.synthetic,

          getPosition: (d) => d.geometry.coordinates as [number, number],
          getFillColor: (d) => {
            return groups.selectionIds.has(d.id as RawId)
              ? WHITE
              : LINE_COLORS_SELECTED_RGB;
          },
          getLineColor: (d) => {
            return groups.selectionIds.has(d.id as RawId)
              ? LINE_COLORS_SELECTED_RGB
              : WHITE;
          },
          getLineWidth: 1.5,
          getRadius: (d) => {
            const id = Number(d.id || 0);
            const fp = d.properties?.fp;
            if (fp) return 10;
            return id % 2 === 0 ? 5 : 3.5;
          },
        }),

        ephemeralState.type === "lasso" &&
          new PolygonLayer<number[]>({
            id: DECK_LASSO_ID,
            data: [makeRectangle(ephemeralState)],
            visible: ephemeralState.type === "lasso",
            pickable: false,
            stroked: true,
            filled: true,
            lineWidthUnits: "pixels",
            getPolygon: (d) => d,
            getFillColor: LASSO_YELLOW,
            getLineColor: LASSO_DARK_YELLOW,
            getLineWidth: 1,
          }),
      ],
    });

    this.lastData = data;
    this.updateSelections(groups.selectionIds);
    this.lastEphemeralState = ephemeralState;
  }

  remove() {
    this.map.remove();
  }

  // Use { diff: false } to force a style load: otherwise
  // if we switch from a style to itself, we don't get
  // a style.load event.
  async setStyle({
    layerConfigs,
    symbolization,
    previewProperty,
  }: {
    layerConfigs: LayerConfigMap;
    symbolization: ISymbolization;
    previewProperty: PreviewProperty;
  }) {
    if (
      layerConfigs === this.lastLayer &&
      symbolization === this.lastSymbolization &&
      previewProperty === this.lastPreviewProperty
    ) {
      return;
    }
    this.lastLayer = layerConfigs;
    this.lastSymbolization = symbolization;
    this.lastPreviewProperty = previewProperty;
    const style = await loadAndAugmentStyle({
      layerConfigs,
      symbolization,
      previewProperty,
    });
    this.map.setStyle(style);

    await new Promise((resolve) => setTimeout(resolve, 100));

    if (this.lastData) {
      this.setData({
        data: this.lastData,
        ephemeralState: this.lastEphemeralState,
        force: true,
      });
      this.lastSelection = { type: "none" };
    }
  }

  private updateSelections(newSet: Set<RawId>) {
    if (!this.map || !(this.map as any).style) return;
    const oldSet = this.lastSelectionIds;
    const tmpSet = new Set(newSet);
    // let adds = 0;
    // let removes = 0;

    // In new set, but not in old set: add to selection
    for (const id of tmpSet) {
      if (!oldSet.has(id)) {
        // If this selection id is a base feature, make all of its
        // vertexes visible
        this.map.setFeatureState(
          {
            source: FEATURES_SOURCE_NAME,
            id,
          },
          {
            state: "selected",
          }
        );
        tmpSet.delete(id);
        // adds++;
      }
    }

    // In old set, but not in new set: remove from selection
    for (const id of oldSet) {
      if (!tmpSet.has(id)) {
        this.map.removeFeatureState(
          {
            source: FEATURES_SOURCE_NAME,
            id,
          },
          "state"
        );
        // removes++;
      }
    }

    // if (adds || removes) {
    //   console.log("adds", adds, "removes", removes);
    // }

    this.lastSelectionIds = newSet;
  }
}
