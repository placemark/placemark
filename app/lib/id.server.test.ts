import { expect, describe, it, test } from "vitest";

import { fcLineString, fcMultiPoint, fcMultiPoly, fcPoly } from "test/helpers";
import type { IFeature, GeometryCollection } from "types";
import { generateSyntheticPoints } from "app/lib/pmap/generate_synthetic_points";
import * as jsonpointer from "app/lib/pointer";
import { Random, MersenneTwister19937 } from "random-js";
import {
  encodeId,
  decodeId,
  encodeFeature,
  encodeVertex,
  nanoid,
  encodeMidpoint,
  idToJSONPointers,
} from "./id";

test("nanoid", () => {
  expect(nanoid()).toHaveLength(21);
});

describe("encode/decode", () => {
  it("#encodeFeature", () => {
    expect(decodeId(encodeFeature(0))).toEqual({
      type: "feature",
      featureId: 0,
    });
  });
  it("#encodeVertex", () => {
    expect(decodeId(encodeVertex(0, 1))).toEqual({
      type: "vertex",
      featureId: 0,
      vertex: 1,
    });
  });
  it("#encodeMidpoint", () => {
    expect(decodeId(encodeMidpoint(0, 1))).toEqual({
      type: "midpoint",
      featureId: 0,
      vertex: 1,
    });
  });
  it("#fuzz", () => {
    const engine = new Random(MersenneTwister19937.seed(1));
    for (let i = 0; i < 100; i++) {
      const input: Id = engine.pick([
        {
          type: "vertex",
          featureId: engine.integer(0, 1000),
          vertex: engine.integer(0, 1000),
        },
        {
          type: "midpoint",
          featureId: engine.integer(0, 1000),
          vertex: engine.integer(0, 1000),
        },
        {
          type: "feature",
          featureId: engine.integer(0, 1000),
        },
      ]);
      expect(decodeId(encodeId(input))).toEqual(input);
    }
  });
});

describe("syntheticIds", () => {
  it("GeometryCollection", () => {
    const gc: IFeature<GeometryCollection> = {
      type: "Feature",
      properties: {},
      geometry: {
        type: "GeometryCollection",
        geometries: [
          fcLineString.features[0].geometry,
          fcMultiPoint.features[0].geometry,
        ],
      },
    };
    const ids = generateSyntheticPoints(gc, 0).map((f) =>
      decodeId(f.id as RawId)
    ) as VertexId[];

    expect(ids.map((id) => idToJSONPointers(id, gc))).toMatchSnapshot();
  });
  it("LineString", () => {
    const ids = generateSyntheticPoints(fcLineString.features[0], 0).map((f) =>
      decodeId(f.id as RawId)
    ) as VertexId[];

    expect(idToJSONPointers(ids[0], fcLineString.features[0])).toEqual([
      "/geometry/coordinates/0",
    ]);
    expect(idToJSONPointers(ids[2], fcLineString.features[0])).toEqual([
      "/geometry/coordinates/1",
    ]);
  });
  it("Polygon", () => {
    const ids = generateSyntheticPoints(fcPoly.features[0], 0).map((f) =>
      decodeId(f.id as RawId)
    ) as VertexId[];

    expect(idToJSONPointers(ids[0], fcPoly.features[0])).toEqual([
      "/geometry/coordinates/0/0",
      "/geometry/coordinates/0/3",
    ]);
  });
  it("MultiPolygon", () => {
    const ids = generateSyntheticPoints(fcMultiPoly.features[0], 0).map((f) =>
      decodeId(f.id as RawId)
    ) as VertexId[];
    const [pointer] = idToJSONPointers(ids[4], fcMultiPoly.features[0]);
    expect(pointer).toEqual("/geometry/coordinates/0/0/2");
    const before = fcMultiPoly.features[0];
    const after = jsonpointer.clone(fcMultiPoly.features[0], pointer);
    expect(before).not.toBe(after);
    expect(before).toEqual(after);
    expect(before).not.toBe(after);
    jsonpointer.set(after, pointer, [90, 90]);
    expect(jsonpointer.get(after, pointer)).toEqual([90, 90]);
    expect(jsonpointer.get(before, pointer)).toEqual([2, 3]);
  });
});
