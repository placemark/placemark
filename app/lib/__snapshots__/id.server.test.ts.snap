// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`syntheticIds > GeometryCollection 1`] = `
[
  [
    "/geometry/geometries/0/coordinates/0",
  ],
  [
    "/geometry/geometries/0/coordinates/0",
  ],
  [
    "/geometry/geometries/0/coordinates/1",
  ],
  [
    "/geometry/geometries/0/coordinates/1",
  ],
  [
    "/geometry/geometries/0/coordinates/2",
  ],
  [
    "/geometry/geometries/1/coordinates/0",
  ],
  [
    "/geometry/geometries/1/coordinates/1",
  ],
]
`;
